pkg/placement/membership.go

```diff
@@ -314,69 +148,87 @@ func (p *Service) processRaftStateCommand(ctx context.Context) {
 			case raft.TableDisseminate:
 				// TableDisseminate will be triggered by disseminateTimer.
 				// This disseminates the latest consistent hashing tables to Dapr runtime.
-				p.performTableDissemination(ctx)
+				p.performTableDissemination(ctx, op.host.Namespace)
 			}
 		}
 	}
 }
 
-func (p *Service) performTableDissemination(ctx context.Context) error {
-	p.streamConnPoolLock.RLock()
-	nStreamConnPool := len(p.streamConnPool)
-	p.streamConnPoolLock.RUnlock()
-	nTargetConns := len(p.raftNode.FSM().State().Members())
+func (p *Service) handleDisconnectedMember(op hostMemberChange, updated bool) bool {
+	p.lastHeartBeat.Delete(op.host.NameAndNamespace())
+
+	// If this is the last host in the namespace, we should:
+	// - remove namespace-specific data structures to prevent memory-leaks
+	// - prevent next dissemination, because there are no more hosts in the namespace
+	if p.raftNode.FSM().State().MemberCountInNamespace(op.host.Namespace) == 0 {
+		p.disseminateLocks.Delete(op.host.Namespace)
+		p.disseminateNextTime.Del(op.host.Namespace)
+		p.memberUpdateCount.Del(op.host.Namespace)
+		updated = false
+	}
+	return updated
+}
+
+func (p *Service) performTableDissemination(ctx context.Context, ns string) error {
+	nStreamConnPool := p.streamConnPool.getStreamCount(ns)
+	if nStreamConnPool == 0 {
+		return nil
+	}
+
+	nTargetConns := p.raftNode.FSM().State().MemberCountInNamespace(ns)
 
-	monitoring.RecordRuntimesCount(nStreamConnPool)
-	monitoring.RecordActorRuntimesCount(nTargetConns)
+	monitoring.RecordRuntimesCount(nStreamConnPool, ns)
+	monitoring.RecordActorRuntimesCount(nTargetConns, ns)
 
 	// Ignore dissemination if there is no member update
-	cnt := p.memberUpdateCount.Load()
+	ac, _ := p.memberUpdateCount.GetOrSet(ns, &atomic.Uint32{})
```

Comment:
Is that the err we are ignoring? I don't see it used anywhere. Is it worth removing, or worth keeping and adding the error check?

Reply:
The ignored value just tells us if there was already an existing key with that name or not. Since we don't need that information here, we can safely ignore it.